### DISCUSSION TITLE
Vote for more storage options

### DISCUSSION THREAD

**Discussion by inian** [supabase/supabase#982] [Feature Requests]
This issue is for tracking requests/demand for different storage backends support in our [storage middleware](https://github.com/supabase/storage-api). Give a comment a thumbs up if you want us to add support for your favourite storage backend.

Please refrain from adding comments unless you are adding a backend which isn't already listed.

**Comment by inian** [comment]
S3 - Supported

**Comment by inian** [comment]
Wasabi

**Reply by fangmarks** [reply]
Wouldn't it already be possible to support Wasabi though?

I mean, it already supports S3, which Wasabi is compatible with.

**Reply by inian** [reply]
Absolutely. I have already exposed an option where you can pass in a non-S3 endpoint [here](https://github.com/supabase/storage-api/blob/78a277ed82de3c350e02c372f87b17a659f9e6b2/src/utils/s3.ts#L25). That being said, I haven't tested it out with an actual Wasabi setup yet. If you can help test it out, that would be great!

**Reply by fangmarks** [reply]
I'd love to test it out but I'm unsure where to start 😅  
I haven't touched [supabase/storage-api](https://github.com/supabase/storage-api) at all yet

**Comment by inian** [comment]
Backblaze

**Reply by sixcorners** [reply]
Just noting here that b2 (backblaze) is s3 compatible... as long as your bucket was created after they announced the feature... Is this proposal for supporting the b2 protocol directly or just making sure it's compatible with b2's s3 api?

**Comment by inian** [comment]
Google Cloud Storage

**Reply by atnegrete** [reply]
This is huge, especially for migrating off firebase

**Reply by arunpkm** [reply]
Please add this support. I need to exhaust my GCP credit :)

**Comment by inian** [comment]
Azure Object Storage

**Reply by dsinclairmoore** [reply]
If this doesn't happen soon, I'll just write my own implementation for this as it's pretty common. 

**Workaround**
If you use container apps and set the container app environment to use Azure files. The storage file path of the app volume can be set as Azure Files (blob store underneath), this is effectively the same thing (with more effort)

**Comment by silentworks** [comment]
Digital Ocean Spaces Object Storage

**Reply by arnu515** [reply]
Digital Ocean Spaces is S3-Compatible, so it will probably work.

**Comment by joshnuss** [comment]
Cloudinary

**Comment by toxsick** [comment]
Minio

**Reply by inian** [reply]
Hi @toxsick, Minio is a storage middleware like our storage-api.

We ended up building our own middleware instead of using Minio. You can read more about our decision in our blog - https://supabase.io/blog/2021/03/30/supabase-storage

**Reply by toxsick** [reply]
Hi @inian, jeah I have seen that post. I guess my question is what selfhosted options will be supported, since I have only seen Paas in the suggestions here. Thanks!

**Comment by JPGallegos1** [comment]
Cloudinary!

**Comment by eugenehp** [comment]
ipfs.com

from here https://www.producthunt.com/posts/supabase-storage#comment-1311900

**Reply by sixcorners** [reply]
How would this work?

**Reply by adelin-b** [reply]
Here is a db using ipfs under the hood
https://github.com/orbitdb/orbit-db

**Reply by angelorc** [reply]
I had the same idea. I'm working on a version that integrates IPFS as storage.

**Comment by 0xflotus** [comment]


**Reply by silentworks** [reply]
Yes, this was outlined in the article on storage https://supabase.io/blog/2021/03/30/supabase-storage#integration-with-the-supabase-ecosystem

**Reply by 0xflotus** [reply]
What the heck? I had read the article and did not recognised it. 🙈

**Comment by fangmarks** [comment]
[Bunny](https://bunny.net)

**Reply by agencyhub** [reply]
+1 for Bunny! Incredible speed!

**Comment by ipeenmaa** [comment]
Linode Object Storage (it is also S3 compliant)

**Comment by acejam** [comment]
[Filebase](https://filebase.com/)

**Reply by fangmarks** [reply]
Should already be supported, since it's S3 compatible, the option to change the storage provider just needs to be public, if it isn't already

**Comment by madx** [comment]
[Scaleway Object Storage](https://www.scaleway.com/en/object-storage/) (S3-compatible)

**Comment by jacobwgillespie** [comment]
[Cloudflare R2](https://blog.cloudflare.com/introducing-r2-object-storage/)

It's not publicly launched yet, but would be an amazing integration once it is. Waitlist access can be found [here](https://www.cloudflare.com/r2-storage/), you might be able to get early access to develop an integration.

**Reply by LyMc** [reply]
Would it be possible to integrate with [Cloudflare images](https://www.cloudflare.com/products/cloudflare-images)?

**Reply by uxduck** [reply]
[Cloudflare R2](https://www.cloudflare.com/en-gb/products/r2/) implements an "S3-compatible API", so this might be relatively simple compared to the other requests?

**Reply by inian** [reply]
AFAIK it is only accessible from inside Workers environment

**Reply by riderx** [reply]
The service is public now, and get be a better placeholder than S3 for Supabase cloud also since egress is free and storage is at the edge

**Reply by texastoland** [reply]
[S3 API Compatibility](https://developers.cloudflare.com/r2/api/s3/api/) for the team's reference 🤞🏼

**Reply by samuelbohl** [reply]
Looks like there is finally an open PR that includes also R2 compatibility: https://github.com/supabase/storage-api/pull/350
Thanks @fenos

**Comment by bhvngt** [comment]
@inian Does supabase currently supports switching the backend storage from S3 to backblaze or any other S3 compatible storage or is this yet to be rolled out? If it does support then is there any doc available to configure supabase setup to change the default storage (I believe it is currently S3).

**Reply by inian** [reply]
For the hosted version, only s3 is supported. If you are self hosting, technically any S3 compliant storage should already be compatible (we don't have tests for this though). If you are self hosting with a storage other than s3, you can [set the](https://github.com/supabase/storage-api/blob/9480891af024396c58045578d16c91778aae67d2/src/utils/config.ts#L55) `GLOBAL_S3_ENDPOINT` variable accordingly.

**Reply by bhvngt** [reply]
Thanks @inian. Is there a plan to support non-S3 (but S3 compatible) backend for hosted version?

**Reply by inian** [reply]
Not in the immediate future, may I ask why you are looking for another provider in the hosted version? The provider we use in the backend is managed and something you won't have direct access to anyway.

**Reply by bhvngt** [reply]
Ask for different providers was because of the egress cost. One of our use case is egress cost sensitive. Hence, I was checking if supabase has plans to let the users of their hosted platform choose providers with more favourable costing structure. Mu understanding is that the current supabase egress cost for storage is relatively high due to S3's high egress cost.

**Reply by baikjy1004** [reply]
S3 storage price and egress fee are expensive. Not only Backblaze B2 but also other object storage providers offer much lower costs, especially for the egress fee. Even without CDN, their egress fee is much cheaper. I would say Supabase hosted version needs to provide a competitive price for the object storage to make the developers use their hosted version more.

**Reply by inian** [reply]
The problem is that since the storage API is running in AWS, we would be paying those expensive egress costs anyway. The request looks like this - user <-> CDN <-> AWS <-> S3. Currently the main egress cost for us is cost out of AWS to the CDN. Switching from S3 to B2 actually would cost more since downloading from S3 to our API servers is actually free but downloading from B2 to AWS would have B2's egress fees. So there is no cost savings we can pass to our users even if we make the switch.

**Reply by bhvngt** [reply]
This is understandable @inian. In some sense, AWS has sort of a lock-in eco system when comes egress fees. Since they are not part of [Bandwidth Alliance](https://www.cloudflare.com/en-gb/bandwidth-alliance/), any data coming out of AWS eco system to anywhere gets charged at their standard egress cost. This becomes a deal breaker for storage heavy use cases such as media files as cost difference is substantial. Unless there is a way to host API server to more egress friendly cloud provider.

**Reply by inian** [reply]
We are planning to offer hosted Supabase in GCP as well at some point. The bandwidth alliance will become more relevant there and after that, integrating with more storage providers like B2 would help lower our costs

**Reply by bhvngt** [reply]
Thats great to hear @inian. Will be looking forward to that news whenever it happens.

**Reply by riderx** [reply]
@inian thanks for the info, I believe allow switch to a storage who implements S3 is achievable in few days. at least at project creation.
Who could already handle some of our needs

**Reply by riderx** [reply]
@bhvngt it seems GCP is not doing the free egress anymore.

**Reply by bhvngt** [reply]
Ohh.. Thanks for that info @riderx. In any case there are alternative such as [cloudflare R2](https://www.cloudflare.com/pg-lp/r2) which don't charge anything for egress and have `S3 compatible` API. 

It will be great if supabase can make it easy to use other S3 compatible alternatives on their hosted platform. From what I understand, they do support other such platform through a [config](https://github.com/supabase/storage-api/blob/9480891af024396c58045578d16c91778aae67d2/src/utils/config.ts#L55). However,  this configuration is currently not supported on their hosted platform.

**Comment by HosamHasanRamadan** [comment]
Contabo  [object storage](https://contabo.com/en/object-storage/)

**Comment by sebastiangraz** [comment]
[Hetzner](https://www.hetzner.com/ )

**Comment by GiantappMan** [comment]
[TencentCloud COS](https://www.tencentcloud.com/products/cos)

**Comment by meotimdihia** [comment]
[seaweedfs](https://github.com/seaweedfs/seaweedfs)

Optimized for small files.

![image](https://user-images.githubusercontent.com/300961/196205499-eb3a36ee-2e08-4967-a7e9-6d96ad321941.png)

**Comment by antran22** [comment]
FTP/rsync

**Comment by jtkeyva** [comment]
Is backblaze an option yet?

**Comment by Zegoverno** [comment]
Can we add a FTP server to upload and download files to the supabase?

**Comment by kmozurkewich** [comment]
Storj (storj.io)

**Comment by jtkeyva** [comment]
Back Blaze B2
